client/src/portal/mod.rs

```diff
@@ -374,12 +374,61 @@ impl Keyring {
     }
 
     /// Return key, derive and store it first if not initialized
-    async fn derive_key(&self) -> &Key {
-        if self.key.get().is_none() {
-            let key = self.keyring.read().await.derive_key(&self.secret);
-            self.key.set(key).unwrap();
+    async fn derive_key(&self) -> Arc<Key> {
+        let keyring = Arc::clone(&self.keyring);
+        let secret = Arc::clone(&self.secret);
+
+        let mut key_lock = self.key.lock().await;
+        if key_lock.is_none() {
+            #[cfg(feature = "async-std")]
+            let key = blocking::unblock(move || {
+                async_io::block_on(async { keyring.read().await.derive_key(&secret) })
+            })
+            .await;
+            #[cfg(feature = "tokio")]
+            let key =
+                tokio::task::spawn_blocking(move || keyring.blocking_read().derive_key(&secret))
+                    .await
+                    .unwrap();
+
+            *key_lock = Some(Arc::new(key));
         }
-        self.key.get().unwrap()
+
+        Arc::clone(key_lock.as_ref().unwrap())
+    }
+
+    /// Change keyring secret
+    ///
+    /// # Arguments
+    ///
+    /// * `secret` - The new secret to store.
+    pub async fn change_secret(&mut self, secret: Secret) -> Result<(), Error> {
+        #[cfg(feature = "tracing")]
+        tracing::debug!("Changing keyring secret and key");
+
+        let keyring = self.keyring.read().await;
+        let key = self.derive_key().await;
+
+        let mut items = Vec::new();
+        for item in &keyring.items {
+            items.push(item.clone().decrypt(&key)?);
+        }
+        drop(keyring);
+
+        self.secret = Arc::new(secret);
+        self.key = Default::default();
```

Comment:
```suggestion
        // Unset the old key
        *self.key.get_mut().unwrap() = None;
```

Default::default works but is not clear what it does in this case

Reply:
yeah, make sense. done.